### PR TITLE
✨ Watch dockercluster

### LINKED ISSUE #223: Add Watch for Cluster/DockerCluster in machine controller
/kind feature

**Describe the solution you'd like**
ATM machine controller requeue with arbitrary delay when waiting for conditions on Cluster/DockerCluster to happen.
This can be improved by adding a watch for Cluster/DockerCluster in machine controller;

**Anything else you would like to add:**
From slack...
> if I got this right, every time Cluster/DockerCluster changes, we should trigger a request for all the DockerMachine belonging to the cluster, right?
> ah, watch DockerCluster => queue DockerMachines based on DockerCluster. you will probably have to write your own MapTo function instead of using the Util since one DockerCluster will be associated with multiple DockerMachines

### PR DESCRIPTION
**Pull request by chuckha**
**What this PR does / why we need it**:
This PR Watches DockerClusters and requeues all DockerMachines when DockerCluster updates.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #223

### CODE REVIEW

File: controllers/dockermachine_controller.go [hunk 1]
```diff
@@ -243,5 +241,48 @@ func (r *DockerMachineReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+func (r *DockerMachineReconciler) DockerClusterToDockerMachines(o handler.MapObject) []ctrl.Request {
+	result := []ctrl.Request{}
+	c, ok := o.Object.(*infrav1.DockerCluster)
+	if !ok {
+		r.Log.Error(errors.Errorf("expected a DockerCluster but got a %T", o.Object), "failed to get DockerMachine for DockerCluster")
+		return nil
+	}
+	log := r.Log.WithValues("DockerCluster", c.Name, "Namespace", c.Namespace)
+
+	cluster, err := util.GetOwnerCluster(context.Background(), r.Client, c.ObjectMeta)
```
Comment (ncdc) [inline controllers/dockermachine_controller.go]: I think technically `context.TODO()` is supposed to be used if you need a context but don't have one passed in to this function. From what I recall, `context.Background()` really should only be created in `main()`. (I realize we are in no way consistent about this across the CAP* repos).

Comment (chuckha) [reply controllers/dockermachine_controller.go]: totally fair. I noticed the CAPA implementation also uses .Background, but I'll update this

File: controllers/dockermachine_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	c := fake.NewFakeClientWithScheme(setupScheme(), objects...)
+	r := DockerMachineReconciler{
+		Client: c,
+		Log:    klogr.New(),
+	}
+	mo := handler.MapObject{
+		Object: dockerCluster,
+	}
+	out := r.DockerClusterToDockerMachines(mo)
+	if len(out) != 2 {
```
Comment (ncdc) [inline controllers/dockermachine_controller_test.go]: Is there value in checking that you got the correct keys by name?

Comment (chuckha) [reply controllers/dockermachine_controller_test.go]: I'd be surprised if they were duplicated but this is an easy enough check! I'd rather have a test for it here then some weird bug later when I'm working on something else

Comment (ncdc) [reply controllers/dockermachine_controller_test.go]: 👍 I tend to get leery when unit tests check expected lengths where the number is 2 or greater when the specific entries in the list matter

Comment (chuckha) [reply controllers/dockermachine_controller_test.go]: fixed!

File: controllers/dockermachine_controller.go (hunk 1 above)
Comment (ncdc) [reply controllers/dockermachine_controller.go]: missed this one

Comment (chuckha) [reply controllers/dockermachine_controller.go]: fixed it!

**Pull request merged**